File: agent/crypto/src/envelope.cpp
#include "jd_relay/crypto/envelope.h"
#include <stdexcept>

namespace jd_relay::crypto {

const char* to_string(MessageType t) {
    switch (t) {
        case MessageType::BUILD_TRIGGER:        return "BUILD_TRIGGER";
        case MessageType::BUILD_RESULT:         return "BUILD_RESULT";
        case MessageType::SENSITIVE_REVIEW_REQ: return "SENSITIVE_REVIEW_REQ";
        case MessageType::SECOND_REVIEW_RESULT: return "SECOND_REVIEW_RESULT";
        case MessageType::HEARTBEAT:            return "HEARTBEAT";
        case MessageType::ACK:                  return "ACK";
    }
    return "UNKNOWN";
}

MessageType parse_message_type(const std::string& s) {
    if (s == "BUILD_TRIGGER")        return MessageType::BUILD_TRIGGER;
    if (s == "BUILD_RESULT")         return MessageType::BUILD_RESULT;
    if (s == "SENSITIVE_REVIEW_REQ") return MessageType::SENSITIVE_REVIEW_REQ;
    if (s == "SECOND_REVIEW_RESULT") return MessageType::SECOND_REVIEW_RESULT;
    if (s == "HEARTBEAT")            return MessageType::HEARTBEAT;
    if (s == "ACK")                  return MessageType::ACK;
    throw std::invalid_argument("Unknown message type: " + s);
}

std::string build_signing_payload(const CryptoEnvelope& env) {
    // Format: msg_id|timestamp|nonce|type|iv|ciphertext|tag
    return env.msg_id + "|"
         + std::to_string(env.timestamp) + "|"
         + env.nonce + "|"
         + env.type + "|"
         + env.iv + "|"
         + env.ciphertext + "|"
         + env.tag;
}

} // namespace jd_relay::crypto

File: forwarder/app/crypto/envelope.py
"""CryptoEnvelope and MessageType — wire-format identical to C++ jd_relay::crypto.

JSON field names: msg_id, timestamp, nonce, type, iv, ciphertext, tag, signature
Signing payload: msg_id|timestamp|nonce|type|iv|ciphertext|tag
"""

from dataclasses import dataclass, field
from enum import Enum


class MessageType(str, Enum):
    """Message types for the Forwarder ↔ Agent encrypted channel.
    Values match C++ to_string() output exactly."""
    BUILD_TRIGGER = "BUILD_TRIGGER"
    BUILD_RESULT = "BUILD_RESULT"
    SENSITIVE_REVIEW_REQ = "SENSITIVE_REVIEW_REQ"
    SECOND_REVIEW_RESULT = "SECOND_REVIEW_RESULT"
    HEARTBEAT = "HEARTBEAT"
    ACK = "ACK"

    @classmethod
    def parse(cls, s: str) -> "MessageType":
        """Parse from string. Raises ValueError on unknown type."""
        for member in cls:
            if member.value == s:
                return member
        raise ValueError(f"Unknown message type: {s}")


@dataclass
class CryptoEnvelope:
    """The encrypted envelope structure.
    Wire format is JSON with base64-encoded binary fields.
    Field semantics match C++ CryptoEnvelope exactly.
    """
    msg_id: str = ""          # UUID v4
    timestamp: int = 0        # Unix epoch milliseconds
    nonce: str = ""           # base64(16 random bytes)
    type: str = ""            # MessageType string
    iv: str = ""              # base64(12 bytes, AES-GCM nonce)
    ciphertext: str = ""      # base64(AES-256-GCM(plaintext))
    tag: str = ""             # base64(16 bytes, GCM auth tag)
    signature: str = ""       # base64(ECDSA signature over signing payload)

    @property
    def msg_type(self) -> MessageType:
        return MessageType.parse(self.type)


def build_signing_payload(env: CryptoEnvelope) -> str:
    """Build the canonical signing payload.
    Format: msg_id|timestamp|nonce|type|iv|ciphertext|tag
    Matches C++ CryptoEnvelope::build_signing_payload() exactly.
    """
    return "|".join([
        env.msg_id,
        str(env.timestamp),
        env.nonce,
        env.type,
        env.iv,
        env.ciphertext,
        env.tag,
    ])
